tests/app/main/views/test_services.py

```diff
@@ -888,274 +888,230 @@ def _mock_find_audit_events_side_effect(fae_response_src, implicit_page_len, **k
+        },
+    }
+    EXPECTED_OLDEST_EDIT_INFO_3 = "Changed on Saturday 30 June 2012 at 9:01pm"
+    EXPECTED_LATEST_EDIT_INFO_3 = "marion@example.com made 2 edits on Saturday 30 June 2012."
+
+    ENABLED_SERVICE_MULTIPLE_EDITS_BY_SAME_USER = (
+        FIND_AUDIT_EVENTS_API_RESPONSE_3,
+        OLD_VERSION_OF_SERVICE_3,
+        "enabled",
+        EXPECTED_OLDEST_EDIT_INFO_3,
+        ((5, EXPECTED_LATEST_EDIT_INFO_3), (2, EXPECTED_LATEST_EDIT_INFO_3), (1, EXPECTED_LATEST_EDIT_INFO_3))
+    )
+
+    FIND_AUDIT_EVENTS_API_RESPONSE_4 = (
+        {
+            "id": 556677,
+            "type": "update_service",
+            "acknowledged": False,
+            "data": {
+                "oldArchivedServiceId": "3535",
+                "newArchivedServiceId": "6767",
+            },
+            "createdAt": "2005-11-12T15:01:12.345Z",
+            "user": "private.carr@example.com",
+        },
+        {
+            "id": 668833,
+            "type": "update_service",
+            "acknowledged": False,
+            "data": {
+                "oldArchivedServiceId": "6767",
+                "newArchivedServiceId": "7373",
+            },
+            "createdAt": "2005-12-10T11:55:12.345Z",
+            "user": "private.carr@example.com",
+        },
+        {
+            "id": 449966,
+            "type": "update_service",
+            "acknowledged": False,
+            "data": {
+                "oldArchivedServiceId": "7373",
+                "newArchivedServiceId": "4747",
+            },
+            "createdAt": "2005-12-11T12:55:12.345Z",
+            "user": "cissy@example.com",
+        },
+        {
+            "id": 221188,
+            "type": "update_service",
+            "acknowledged": False,
+            "data": {
+                "oldArchivedServiceId": "4747",
+                "newArchivedServiceId": "9292",
+            },
+            "createdAt": "2005-12-17T09:22:12.345Z",
+            "user": "private.carr@example.com",
+        },
+    )
+    OLD_VERSION_OF_SERVICE_4 = {
+        "3535": {
+            "id": "151",
+            "frameworkSlug": "g-cloud-9",
+            "lot": "cloud-hosting",
+            "supplierId": 909090,
+            "supplierName": "Barrington's",
+            "serviceName": "Lemonflavoured soup",
+        },
+    }
+    EXPECTED_OLDEST_EDIT_INFO_4 = "Changed on Saturday 12 November 2005 at 3:01pm"
+    EXPECTED_LATEST_EDIT_INFO_4 = "The last user to edit this page was private.carr@example.com " \
+        "on Saturday 17 December 2005."
+
+    ENABLED_SERVICE_WITH_MULTIPLE_USER_EDITS = (
+        FIND_AUDIT_EVENTS_API_RESPONSE_4,
+        OLD_VERSION_OF_SERVICE_4,
+        "enabled",
+        EXPECTED_OLDEST_EDIT_INFO_4,
+        ((5, EXPECTED_LATEST_EDIT_INFO_4), (3, EXPECTED_LATEST_EDIT_INFO_4), (2, EXPECTED_LATEST_EDIT_INFO_4))
+    )
+
+    ENABLED_SERVICE_WITH_NO_EDITS = (
+        (),
+        {},
+        "enabled",
+        None,
+        ((5, "0 unapproved edits.",),)
+    )
```

Comment:
If you really think this is a better and more maintainable way of declaring these then I don't know what to say. It might *look* clearer right now as it is now (I don't think so personally, I have to do a bunch of variable chasing to see what references what), but the first time someone comes along to modify/attempt to reuse any of these scenarios it will start to go down a dark path.

Reply:
I worked on this refactor with @katstevens and both of us think it is far more readable in the new version. It also made it much easier for us to split asserting the message about recent edits into a separate test.
I do not understand why would things go down the dark path if someone tried to modify or attempt to reuse any of these scenarios in current setup (it went quite well actually when we did it), but not before. Would you mind explaining?
And what is declarative notation?

Reply:
In general, the learning curve to parse all these nested tuples and figure out what we were actually meant to be testing in each case was very steep. The changes in this PR so far feel like a step in the right direction, without losing any subtleties in the combination of parameters.

Saying that, we could probably make our test data a bit less ALL CAPS now that we've clarified what the scenarios are. I'd be happy with (carefully named!) helper functions that set up the elements of these scenarios, that would hopefully be more reusable.

Re: scenario setups - if a more complex scenario meant this test is no longer fit for purpose, it feels like that test should be separate anyway. I was tempted to split out the empty case anyway as the assertions are very different, but thought that might be change too far for this story 😸 .

Reply:
So previously the declaration of the data structure had a form of *referential transparency*. The structure of the file was a fairly strict tree-like hierarchy and every pair of brackets defined a structure or generator completely contained within it. In comparison, variables-referencing-each-other-all-over-the-place is spaghetti, exposing all kinds of internal details that you don't necessarily need to expose. This makes it less clear what the intent of the definitions are (what the intended *output* of the little variable dance is supposed to be), which are *supposed* to be used in tests, etcetera...

Gone is the encapsulation and ability to double click on a bracket and have your editor highlight the whole relevant section of declaration and know there are *no* side effects.